### ISSUE TITLE
x86

### ISSUE THREAD

**Issue by skl131313** [joakim-noah/android#12]
@joakim-noah 

I'm trying to get the x86 version working for D, it seems there's some trouble with getting it up and running though. It seems specifically to be with TLS and the way they are being created.

I get the following error when I just place `int test` in the module and try to use it. I know that TLS is emulated on Android, but I was wondering how it goes about doing that? Is the LLVM backend just generating the wrong types and relocation types? Maybe you have a better idea than I do, I don't know much about this low level implementation stuff. Not sure why the ARM implementation isn't having problems, from what I've seen there isn't a lot of Android specific code in LDC.

```
direct GOT relocation R_386_GOT32 against `_D4main4testi' without base register can not be used when making a shared object
```

Basically only thing I've changed relative to the armv7 build is thie parameter passed to LDC2 to build the D source:

```
ldc2 -mtriple=i686-none-linux-android ...
```

Runtime library for LDC2 builds, but there's the same problem when linking. Get that error for any TLS variables.

Idk if you want to move to email or something else, it's a shame you can't just message a user on their profile..

**Comment by joakim-noah** [comment]
Before we look at the technical issues, can you explain why you want to get Android/x86 working in the first place?  As I mentioned before, this platform is basically dead.

**Comment by skl131313** [comment]
For use with the emulator, it may be dead but pretty much every development PC is x86. Running the ARM emulator is pretty much useless.

**Comment by joakim-noah** [comment]
Correct me if I'm wrong, as I never used the Android SDK much, but don't most Android devs test in an Android/ARM emulator?  That is the most common deployment platform after all.  When I first started the Android port, I installed Android/x86 in a VM and tested on there, but that wasn't what the SDK said to do.

**Comment by skl131313** [comment]
It runs way too slow, the only reason I can think of that ARM would be tested is that it use to be the only option. For me it can take 10+ mins just to boot the ARM emulator, let alone testing an app. If you are only testing the logic of your program, then it doesn't matter which arch you use. You are bound to get better testing on a device for compatibility than an emulator anyways.

**Comment by joakim-noah** [comment]
Yes, that is why I only test on Android/ARM devices these days. If you want to get Android/x86 working, most of the support should already be there, but something likely regressed with [the emulated TLS support for x86](https://github.com/termux/termux-packages/pull/1323#issuecomment-325963476).

If you're unfamiliar with these low-level details, I suggest you simply test on an Android/ARM device, rather than wasting time with x86 support.  [The Termux app](https://play.google.com/store/apps/details?id=com.termux) provides a nice shell where you can run command-line test binaries easily, it even has an ldc package.  [You can setup adb over wifi](http://codetheory.in/android-debug-bridge-adb-wireless-debugging-over-wi-fi/) and quickly push an apk to your device, then actually interact with the touchscreen easily.

As for your ARM issue with zlib, it's not clear why you're loading a library, as Phobos comes with a built-in zlib.  I'd need more info on how you're building your libmain.so.

**Comment by skl131313** [comment]
Fair enough,  it'd probably be too much work.

I am using the zLib in photos,  or rather I don't need it but it is still
linked in. I'm following all the instructions the same way as on the wiki.

**Comment by joakim-noah** [comment]
It's not too much work for me: I just haven't bothered because almost nobody is running that platform and it's better to test with real hardware, which is going to be ARM or at most ARM64 these days.  Another issue for me is that I haven't had an x86/x64 device in a couple years, since my Win7 ultrabook conked out in late 2015, only using ARM these days.  If you're unfamiliar with the low-level implementation details of TLS, [like I was before I started on the Android port a couple years ago](http://forum.dlang.org/thread/ydkehepmzwvtvardilye@forum.dlang.org), it might take you a while.

As for the zlib problem, how exactly does it manifest itself: the app crashes at some point?  How do you know the problem is text relocations: do you see that error in the adb log?  It's possible some bug has crept in, as I haven't been running all the stdlib tests in an apk lately, only from the command-line.

When you call writeln, presumably from a GUI app, where do you expect it to write to?  If you're trying to write to stdout, that's not possible for an Android GUI app: [you have to write to a file](https://gist.github.com/joakim-noah/348edc378d47fb90e32708be19286a2e#file-druntime_1-3_ldc_arm-L80) or to [Android's debug logger by using its `__android_log_`* functions](https://github.com/joakim-noah/android/blob/master/samples/native-activity/jni/main.d#L28).

**Comment by joakim-noah** [comment]
Another thing you can try is adding the flags `-Wl,--warn-shared-textrel -Wl,--fatal-warnings` when linking the shared library.  The NDK adds those and several other flags when building a native shared library for a GUI apk, and [ldc uses them when linking a command-line stdlib test runner too](https://github.com/ldc-developers/ldc/blob/master/runtime/PresetRuntimeConfiguration.cmake#L37), but I removed them from the wiki examples to make them shorter.

**Comment by skl131313** [comment]
The problem is just a warning in the linker but it does treat it as an
error,  if you bypass that though when you try to dlopen the shared library
it says it can't as it has text relocations. I'm using LDC.-1.7.0 and
Android app v24. I ended up just removing the he few files from the D
runtime that used zLib and everything works now,  but there is definitely
something wrong specifically with it, not sure what but it seems text
relocations are related to libraries being loaded at a static address.

**Comment by joakim-noah** [comment]
Yeah, that's why I asked you earlier why you were loading a library.  I'm still not sure what shared library you're trying to dlopen from your code, you're going to have to tell me that before we can go any further.

As for text relocations, they're disallowed since API 23, so they would cause a problem.  Since it's not clear what libraries you're trying to load, need that info first.

**Comment by skl131313** [comment]
It's not from my code, my D code is in a shared library, in that shared library is the static version of the LDC D Runtime. That D runtime contains the zlib library which is also statically linked into the shared library as a result. This zlib that's linked into my shared library is what's causing the issue. The dlopen is called as a result of the Java code `System.loadLibrary` to load a shared library. Don't know if it's possible to create an Android app with a native entry point.

**Comment by joakim-noah** [comment]
Ah, ok, when you said you dlopen a shared library, I thought you were opening some other shared library from your D code.  As the wiki page notes near the end, you can only build a single D shared library right now, you can't have two, which is why I was asking about that.  As I said, it's possible that there's been some regression in building zlib as part of Phobos, I've kicked off a build to investigate it.

And of course you can have a native entry point, that's actually what I assumed, as both the sample apks shown in the wiki page do that, both the first purely native one and the second one that calls some Java methods.

Some of the other sample apps in this repo do demonstrate calling D from a Java entry point, but you have to be careful to initialize the runtime for those, as detailed in [this section of the wiki page](https://wiki.dlang.org/Build_D_for_Android#Changes_for_Android) and [shown with](https://github.com/joakim-noah/android/blob/master/samples/hello-gl2/jni/gl_code.d#L180) [these apps](https://github.com/joakim-noah/android/blob/master/samples/bitmap-plasma/jni/plasma.d#L357).

**Comment by skl131313** [comment]
Alright, thanks for looking into it. Keep me updated, I'm curious as to what is causing it.

I can't find the example for the native binary, it looks like they are all created as shared libraries? https://github.com/joakim-noah/android/blob/master/samples/native-activity/jni/Android.mk#L24

**Comment by joakim-noah** [comment]
You may be right: the test runner apk crashes when checking for certain modules' tests in a shared library, even though it all works fine when run on the Termux command-line.  Not sure if it's related to `std.zlib` though, seems to be failing when looking for various modules' tests.  I'll track it down.

As for a "native binary," I was talking about what you initially mentioned instead, "an Android app with a native entry point."  Any Android app that calls native code in a shared library can first be started from a Java entry point then call into functions from the native shared library, written in D in this case, or tell the Java runtime to call directly into an `android_main` native entry point in the shared library, and then call Java methods if needed.  That's what I was talking about above, the difference between those two and which sample apps demonstrate each type of entry point.

If you want to just build a native command-line binary executable, the sieve example in the wiki shows that.  As for the Makefile you linked, those are unused by this repo for D code.  They're simply there if you want to build the previous C/C++ versions of these sample apps and compare to the D version.  You can't use any of the Makefiles in this repo to build D code.

**Comment by skl131313** [comment]
If you run the following command what output do you get with the shared library that has the d runtime in it? It should show all the text relocations. If it does load then it might not be the same problem, as this prevents loading the library all together. Unless you are running an older version of Android?

```
scanelf -qT ./libmain.so
  libmain.so: $a.1507 [0x2D5FE4] in (optimized out: previous compress_block) [0x2D5CDD]
  libmain.so: $a.1507 [0x2D6094] in (optimized out: previous _tr_tally) [0x2D5FFD]
  libmain.so: $a.83 [0x2D5FF0] in (optimized out: previous $d) [0x2D5FE4]
  libmain.so: $a.83 [0x2D6098] in (optimized out: previous $d) [0x2D6094]
  libmain.so: $d.19 [0x2D65A4] in (optimized out: previous zError) [0x2D6595]
```

Ah yah, that's what I was wondering, so it is possible to create a program that enters through `main()`.

**Comment by joakim-noah** [comment]
I get no output when running your `scanelf` command on a test runner library with all the stdlib modules and it always loads, but sometimes crashes when trying to run certain modules' tests.  I'm testing on Android 6.0, doubt that makes a difference, but which version of Android are you trying?  I suspect your issue might be an artifact of not adding all the linker flags that I use even for the command-line test runner, and which I linked you above.  Try linking with those flags included and see if it makes the difference.

As for entering through `main()`, thats only possible for a command-line executable.  As noted above, you can enter through `android_main` for a shared library, if you use the translated D version of the native glue wrapper that normally comes with the Android NDK.  See some of the C/C++ sample files in this repo to see how that function is called and compare with the D translated version.

**Comment by skl131313** [comment]
Yah I was trying Android 7.0, the issue I'm having didn't happen in early versions of Android. They were just warnings, but they turned that warning into an Error. I think it might be an error still in Android 6.0 though.

If you can upload your druntime, I'll try it, maybe something else is different.

Also do you do debugging with Android studio? Do you know if any plugin exists that'll let you use breakpoints for D code.

**Comment by joakim-noah** [comment]
Hmm, Android 7.0, [are you sure you're running a 32-bit ARM kernel](https://wiki.dlang.org/Build_D_for_Android#Native_compilation)?  If so, did you try adding all the linker flags I mentioned above?

Never used Android Studio, used to attach gdb remotely from my ultrabook, before I switched over to natively developing on Android/ARM.  I don't know of any Studio plugins, but if it's just using gdb, D demangling should be built in.

**Comment by skl131313** [comment]
It's building a 32-bit application yah, i don't think it's the compiler flags. Unless specifically for zlib. Everything else works, once I remove zlib from the druntime. You can't upload your static library of druntime/phobos? It'd knock some things off the list if it does link for me fine.

IF I can ask, what are you developing on?

**Comment by joakim-noah** [comment]
I didn't ask what you're building, I asked whether the kernel you're testing with is 64-bit or 32-bit.  See the link I gave you to figure that out.  As for the linker flags, they will change the shared library generated, so I suggest you try using all of them, as that's what the Android NDK does.  I don't see any point in uploading my libraries till I get some answers to these basic questions first.

I usually develop on an Android/ARM tablet.  When building the shared library for the test runner apk mentioned above, I cross-compile from a linux/x64 VPS, to make sure I'm using a similar setup to most devs who are cross-compiling.

**Comment by skl131313** [comment]
I don't see what difference the the kernel bit version makes, as it doesn't even compile if the default warnings are left on. It could be going to any device and it'd still have the same problem. Yes all the flags have been tried anyways. It's not that difficult to take 2 seconds to drag and drop 2 files that you already have built...

**Comment by joakim-noah** [comment]
What default warning flags did you change to get it to compile?  The reason the kernel matters is because 64-bit is unsupported and you said the problem was "when you try to dlopen the shared library it says it can't as it has text relocations," which is a runtime problem, not just compile-time like you're saying now.

Can you paste "all the flags" you tried here?

It takes a lot longer than "2 seconds" to upload a large test library, which I'm certainly not going to do if you can't be bothered to answer some simple questions first, whose answers could make trying the library useless.

**Comment by skl131313** [comment]
It only becomes a runtime error cause the compile time error is forcibly overridden. It's not like everything stops working either. All I had to do to get it working was remove zlib from phobos. If it was a problem with flags or other compiler flags will my compilations of the shared library then removing a small part of the library wouldn't have solved the problem. 

The number was metaphorical, it wasn't to be taken literally. The questions would be better answered if there was more information to go on. We can sit here all day and just ask questions, that might not even be related to the problem at hand. Asking whether the device is 32-bit or 64-bit, is one such question. It's a step backwards, not forwards. Anyways the library isn't that big, I started uploading this 10mb file at the start of me writing this and it finished long before I finished writing this message.
[somefile.txt](https://github.com/joakim-noah/android/files/1705448/somefile.txt)

Of course! You are free to do what you want, the fact you want me to answer some questions rather than allowing sending a file to diagnose the issue further with more information by comparing the difference between our two libraries, gives me insight into the type of individual you are. The fact you keep repeatedly telling me that the problem are the compiler flags and about the link you gave me, yes I checked all those when you first posted them. I had already double checked them before even posting here. The fact you ignored my first request to upload the libraries all together as well. The fact you consider 20mb to be a "large" file. I can surmise that you will probably not upload the files. It's fine if your upload speed is as fast as dial-up, the files aren't necessary to fix the problem. I'll spend some more time looking into it when I actually have time but I don't really need it was everything is working perfectly except for the compilation of the zlib. It was mostly just curiosity and maybe fix LDC up so that a workaround isn't necessary. Thanks for the help.

**Comment by joakim-noah** [comment]
I could rebut every point you've made here, but I'm not going to bother.  You repeatedly get details wrong throughout this thread- for example, you keep mentioning compiler flags, whereas I pointed you at _linker_ flags- and when I point out what you're saying doesn't make sense or ask you questions to figure out exactly what you're doing, you repeatedly ignore that point or me.

It is clear you have no idea what you're doing, because any time there's an issue like this, you can only get help from someone else by presenting them with the same data you have, which is why I asked you questions.  My sending you a library that works will accomplish nothing, you telling me what flags or device you used for a library that doesn't work is the only way to figure this out.

But since you crazily refuse to do that, it's good you closed this issue, as I would have anyway after this last refusal to actually send me the necessary data.

**Comment by skl131313** [comment]
So you say you aren't going to rebut, but then in the next sentence you go on to argue semantics that isn't any more than a nitpick. I'm glad you didn't try to create an argument against the rest if that is to be presumed that was your best argument, as you were willing to contradict yourself with it. 

>  you telling me what flags or device you used for a library that doesn't work is the only way to figure this out.

Device doesn't matter, tested it on both 32-bit and 64-bit and the result is the same (no surprise) and like I was saying you can see all the command line arguments used from [here](https://wiki.dlang.org/Build_D_for_Android#Build_a_command-line_executable). Now, that isn't very useful information is it? Almost like you already had it to begin with, like I was saying..

> My sending you a library that works will accomplish nothing,

It would:
* Identify where the problem lies, if it was my compiler flags it wouldn't work, if it wasn't then it would work flawless. It would actually have helped your case if it was so.
* Compare the generated binary to identify how the text relocations were removed from your library, so that it could be replicated and applied to remove them from LDC as well.
* See what version of LDC you are actually using, as I have a feeling you aren't compiling the latest one from LDC. Which only makes me feel like something is more off with the amount of push I received from you, in regard with sending the library.

> It is clear you have no idea what you're doing, because any time there's an issue like this, you can only get help from someone else by presenting them with the same data you have

And now you claim to be omniscient, or how do you propose to know about every issue I've had and how I solved them? If you knew what you were doing, since you are certain it is a problem with the flags. You would be able to tell me which flags I am missing that are causing the problem. But you can't cause you are so off base as that isn't even the problem.

> it's good you closed this issue, as I would have anyway after this last refusal to actually send me the necessary data.

This was never meant to be an actual issue. Github simply lacks discussions. I mention this in the very first post.

**Comment by joakim-noah** [comment]
Lol, I said I wouldn't rebut _every_ point, hilarious how you can't read.  Linker flags, not compiler flags as you again wrongly state, is not a matter of "semantics:" it's an important technical detail to getting the build right.  Compiler flags control code generation, while linker flags merely control how the final shared library is packed together.

Depending on which is causing the problem, you go down completely different technical and debugging paths, but of course, these are all mere "nitpicks" for you to hand-wave away.  The compiler and linker are supposed to magically divine what you want, and so am I.

It was important info to know you tested on both 32-bit and 64-bit devices, yet you loonily decided to keep that data from me, after I repeatedly asked for it.

You are clearly nuts and incompetent at these kinds of technical topics, so there is no point in engaging with you further.

**Comment by skl131313** [comment]
Sigh, you can't even see that it is just semantics. They are so closely related only a fool would nitpick whether what one is called. Of course I'm going to continue to call them that, why wouldn't I? It doesn't make a difference, you are just so anal about the tiniest insignificant detail.

```
ldc2 Xcc=-fpie
[Compiler "ldc2"] [Compiler Flags ;)]
```

Like I said semantics.

> It was important info to know you tested on both 32-bit and 64-bit devices, yet you loonily decided to keep that data from me, after I repeatedly asked for it.

No it isn't, like I said I tested it on both, surprise surprise, they both have the same issue. You gain NOTHING from knowing that information. Lol and you still think it's important, how sad.

> You are clearly nuts and incompetent at these kinds of technical topics, so there is no point in engaging with you further.

So you retort to name calling to name calling rather than creating a counter argument for my reasoning as to why you providing your LDC libraries would have helped solve the problem better than me telling you useless information?

**Comment by joakim-noah** [comment]
I'm not going to argue with an incompetent nutcase.  Good luck with your Android experiments, as you will need a lot of luck if you're ever to coincidentally come up with a working app, since you're clearly incapable of thinking clearly and rationally about building something.

**Comment by skl131313** [comment]
And on with the name calling instead of providing an actual argument. Having a different matter of opinion is fine and disagreeing about that difference is only natural. But I think it shows something about that individuals character when the only thing they are posting is to name call someone else.

I have provided you with the information you asked for, you could easily have proved me incorrect by solving the issue with the only details you said you required that was "the only way" to solve the problem. As you are unable to continue on with an intellectual discussion there is no reason to further this discussion with you.

**Comment by ParticlePeter** [comment]
LOL, this issue made my day :-). First, NO INSULT INTENDED at all, but please laugh with me if you can, this here looks hilarious from the outside. I'll explain bellow, but first what brought me here:

I was following your [announcement](https://forum.dlang.org/thread/antajtnvmavswjvcdoyq@forum.dlang.org) and [wiki endtry](https://wiki.dlang.org/Build_D_for_Android), but I couldn't find what I am searching for. I found this repo and searched it for dlopen, and this issue/thread was the only result.

I maintain DLang Vulkan binding [ErupteD](https://github.com/ParticlePeter/ErupteD), it recently got a lib loading mechanism for [Windows and Posix](https://github.com/ParticlePeter/ErupteD/blob/86d2b8e741e2813898a23eb4d99d0ce73c1ad65a/source/erupted/vulkan_lib_loader.d#L16). I would like to add Android to the mixe and figured that, if anybody knows how to get this done, it would be you. I know the proper commands for Android C, but would not know what needs to be imported to get at Android variants of `dlopen`, `dlsym` and `dlclose`. Can you help me out with this?

Sorry to `misuse` this issue for my purpose, didn't want to open a new issue and wouldn't know how else to contact you.

To the fun part :-)
So, this issue, as it is displayed for me at least, starts with skl131313 post:
 > Name calling owner.

what ever that means. This is followed with 15 comments from you, kind of talking to your self and slowly heating up. It ends with your comment ... well yeah, see yourself above this one :-)

I am sure that there was a dialog between you and skl131313, but its gone, without github informing that he deleted some messages. Again, no insult intended!

**Comment by joakim-noah** [comment]
Yes, it does, everything but `Windows`.

**Comment by ParticlePeter** [comment]
Thanks, hard to find that info.

**Comment by joakim-noah** [comment]
If you ever want to check what versions are active, add the `-v` verbose flag to your compilation command and pipe the output to `more`. All the D compilers should list the versions used: `dmd -v test.d | more`
